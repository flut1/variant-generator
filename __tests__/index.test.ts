import createVariants, { iterateCombinations } from '../src/index';

describe('createVariants()', () => {
    test('handles simple 2x2 input', () => {
        const result = createVariants(function*() {
            return {
                bool: yield [false, true],
                num: yield [0, 1],
            };
        });

        expect(Array.from(result)).toMatchSnapshot();
    });
    test('handles simple 3x2 input', () => {
        const result = createVariants(function*() {
            return [yield [0, 1], yield [2, 3], yield [4, 5]];
        });

        expect(Array.from(result)).toMatchSnapshot();
    });
    test('works with a generator without yields', () => {
        const result = createVariants(function*() {
            return [1, 2, 3];
        });

        expect(Array.from(result)).toMatchSnapshot();
    });
    test('creates different variations for deep-equal inputs', () => {
        const result = createVariants(function*() {
            return [yield [0, 1, 2], yield [0, 1, 2]];
        });

        expect(Array.from(result)).toMatchSnapshot();
    });
    test('throws on yields with different lengths', () => {
        expect(() => {
            const inputs = [[1, 2, 3], [1, 2, 3, 4]];
            let pickInputAtIndex = 0;

            function getInput() {
                pickInputAtIndex = pickInputAtIndex ? 0 : 1;
                return inputs[pickInputAtIndex];
            }

            const resultIterator = createVariants(function*() {
                return [yield getInput()];
            });

            // iterate limited number of times to prevent infinite loop
            resultIterator.next();
            resultIterator.next();
            resultIterator.next();
            resultIterator.next();
            resultIterator.next();
        }).toThrowErrorMatchingSnapshot();
    });
    test('throws on yields with different types', () => {
        expect(() => {
            const inputs = [[1, 2, 3], [1, 2, 'foo']];
            let pickInputAtIndex = 0;

            function getInput() {
                pickInputAtIndex = pickInputAtIndex ? 0 : 1;
                return inputs[pickInputAtIndex];
            }

            const resultIterator = createVariants(function*() {
                return [yield getInput()];
            });

            // iterate limited number of times to prevent infinite loop
            resultIterator.next();
            resultIterator.next();
            resultIterator.next();
            resultIterator.next();
            resultIterator.next();
        }).toThrowErrorMatchingSnapshot();
    });
    test(`doesn't duplicate variations for strictly equal inputs`, () => {
        const input = [0, 1, 2];
        const result = createVariants(function*() {
            return [yield input, yield input];
        });

        expect(Array.from(result)).toMatchSnapshot();
    });
    test('throws on a yield with a non-array', () => {
        expect(() => {
            const test = Array.from(
                createVariants(function*() {
                    return yield { 0: 'foo', 1: 'bar' };
                }),
            );
        }).toThrowErrorMatchingSnapshot();
    });
    test('handles conditional yields', () => {
        const result = createVariants(function*() {
            const fruit = yield ['banana', 'apple', 'strawberry'];
            let size = 'normal';

            switch (fruit) {
                case 'banana':
                    size = yield ['small', 'normal', 'large'];
                    break;
                case 'apple':
                    size = yield ['small', 'normal'];
                    break;
            }
            let color: string = 'red';

            switch (fruit) {
                case 'banana':
                    color = yield ['yellow', 'green'];
                    break;
                case 'apple':
                    color = yield ['green', 'red'];
                    break;
            }

            return {
                fruit,
                size,
                color,
            };
        });

        expect(Array.from(result)).toMatchSnapshot();
    });
});

describe('iterateCombinations', () => {
  test('it runs once for each combination of parameters', () => {
    const spy = jest.fn();

    iterateCombinations(function*() {
      yield [1, 2, 3];

      yield [4, 5, 6];

      yield [7, 8];

      spy();
    });

    expect(spy).toBeCalledTimes(3 * 3 * 2);
  })
});
