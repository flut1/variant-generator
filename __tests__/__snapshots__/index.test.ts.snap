// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createVariants() creates different variations for deep-equal inputs 1`] = `
Array [
  Array [
    0,
    0,
  ],
  Array [
    0,
    1,
  ],
  Array [
    0,
    2,
  ],
  Array [
    1,
    0,
  ],
  Array [
    1,
    1,
  ],
  Array [
    1,
    2,
  ],
  Array [
    2,
    0,
  ],
  Array [
    2,
    1,
  ],
  Array [
    2,
    2,
  ],
]
`;

exports[`createVariants() doesn't duplicate variations for strictly equal inputs 1`] = `
Array [
  Array [
    0,
    0,
  ],
  Array [
    1,
    1,
  ],
  Array [
    2,
    2,
  ],
]
`;

exports[`createVariants() handles conditional yields 1`] = `
Array [
  Object {
    "color": "yellow",
    "fruit": "banana",
    "size": "small",
  },
  Object {
    "color": "green",
    "fruit": "banana",
    "size": "small",
  },
  Object {
    "color": "yellow",
    "fruit": "banana",
    "size": "normal",
  },
  Object {
    "color": "green",
    "fruit": "banana",
    "size": "normal",
  },
  Object {
    "color": "yellow",
    "fruit": "banana",
    "size": "large",
  },
  Object {
    "color": "green",
    "fruit": "banana",
    "size": "large",
  },
  Object {
    "color": "green",
    "fruit": "apple",
    "size": "small",
  },
  Object {
    "color": "red",
    "fruit": "apple",
    "size": "small",
  },
  Object {
    "color": "green",
    "fruit": "apple",
    "size": "normal",
  },
  Object {
    "color": "red",
    "fruit": "apple",
    "size": "normal",
  },
  Object {
    "color": "red",
    "fruit": "strawberry",
    "size": "normal",
  },
]
`;

exports[`createVariants() handles simple 2x2 input 1`] = `
Array [
  Object {
    "bool": false,
    "num": 0,
  },
  Object {
    "bool": false,
    "num": 1,
  },
  Object {
    "bool": true,
    "num": 0,
  },
  Object {
    "bool": true,
    "num": 1,
  },
]
`;

exports[`createVariants() handles simple 3x2 input 1`] = `
Array [
  Array [
    0,
    2,
    4,
  ],
  Array [
    0,
    2,
    5,
  ],
  Array [
    0,
    3,
    4,
  ],
  Array [
    0,
    3,
    5,
  ],
  Array [
    1,
    2,
    4,
  ],
  Array [
    1,
    2,
    5,
  ],
  Array [
    1,
    3,
    4,
  ],
  Array [
    1,
    3,
    5,
  ],
]
`;

exports[`createVariants() throws on a yield with a non-array 1`] = `"yielded values in createVariants() should be Arrays of options. Got object"`;

exports[`createVariants() throws on yields with different lengths 1`] = `
"The generator passed to createVariants() seems to yield non-deterministic input Arrays.
The generator should yield the same inputs on each call. Yield #0 was inconsistent:
inputs previously had length 4, now yields 3"
`;

exports[`createVariants() throws on yields with different types 1`] = `
"The generator passed to createVariants() seems to yield non-deterministic input Arrays.
The generator should yield the same inputs on each call. Yield #0 was inconsistent:
input option 2 was of type string, is now number"
`;

exports[`createVariants() works with a generator without yields 1`] = `
Array [
  Array [
    1,
    2,
    3,
  ],
]
`;
